### ISSUE TITLE
HttpWebRequest.GetRequestStream returns new stream instance on every invocation

### ISSUE THREAD

**Issue by dallmair** [dotnet/runtime#41403]
### Description

Some third-party code we're depending on does this:

```csharp
var request = WebRequest.CreateHttp(url);
request.Method = "POST";

// Problem: Calling request.GetRequestStream() twice returns same stream object in .NET Framework,
// but different objects in .NET Core 3.1.
using (var writer = new StreamWriter(request.GetRequestStream()))
    writer.Write("Hello, world!");
request.GetRequestStream().Close();

request.GetResponse();
```

Of course, there are things to improve here; Unfortunately, this code is not under our control, so we have no way to change it.

Anyway, the code works fine on .NET Framework 4.7.2 as the `GetRequestStream` method returns the same `System.Net.ConnectStream` object on both calls. In contrast, it fails on .NET Core 3.1.400, because the method returns a new instance on every call, which means that the above code just sends empty POST requests.

Expectation: `GetRequestStream` should return the same object on every invocation.

### Configuration

* .NET Core 3.1.400
* Windows 10.0.19041.388
* x64

### Regression?

Yes. As outlined above, this works fine on .NET Framework 4.7.2, but does not work on .NET Core 3.1.

### Other information

Here's the relevant code section, which shows that a new `RequestStream` is created on every invocation: https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L1029-L1056

**Comment by karelz** [comment]
Thanks for your report. Can you please help me understand your situation a bit more?
1. How do you use the 3rd party code on .NET Core? Is it a .NET Framework DLL you are running on .NET Core?
2. Is there option to ask the 3rd party to either change their code or target .NET Standard or .NET Core?
3. How wide-spread is usage of the 3rd party code? Is it exclusive to your project, your company, or is it commonly used library across internet?
4, How much does it block you? Do you have option to work around the problem? (e.g. use different code or library)

For context: `HttpWebRequest` is considered obsolete/legacy API on .NET Core (not performant and not fully featured as on .NET Framework). We recommend users to use `HttpClient` instead.
That said, we fix some incompatabilities if they affect larger number of users and block migration to Core.

**Comment by dallmair** [comment]
Sorry for the delay, busy times over here.

First of all, I definitely agree on usage of `HttpClient`. We've updated all of our own code base to use it which was a massive effort, but is way better. We do not regret this switch at all.

We encountered this compatibility issue with the SharePoint Online CSOM library. The SharePoint team added a .NET Standard 2.0 flavor of their library in June, and we tried it out as we're currently planning our product's future and are toying a few ideas, including the switch to .NET Core. Unfortunately, `HttpWebRequest` is part of CSOM's public API and authentication story, so they're kind of stuck with it.

As this might affect others as well (both other users of CSOM as well as other libraries having the same bug), we wanted to at least report it here as a compatibility issue. There's no need for a fix in 3.1 for us right now. Looking a bit further into the future, we think a fix in .NET 5 would be awesome but it might be a little late in the cycle already, not sure about that. Should be possible to fix it in .NET 6, though.

Hope that answers your questions. If there's anything else we can provide, please don't hesitate to ask.

**Comment by stephentoub** [comment]
@danmosemsft, on multiple issues now I see that we tagged the issue as needs more info, then more info was provided, but the needs more info label wasn't removed, and now the bot is going to automatically close the issue. That's not great.

**Comment by dallmair** [comment]
@stephentoub Thanks for the follow-up regarding the bot, good to know you're also monitoring things.

**Comment by ManickaP** [comment]
Triage: still happening. If we think this is worth fixing, it shouldn't be that hard to do, as we already store the stream in `_requestStream`.
cc @liveans for more insights on `WebRequest`.

**Comment by liveans** [comment]
I agree that this is not hard to fix, only thing we need to do is change assignments on `_requestStream` (`=` to `??=`). I also believe this might be a `good first issue`.

**Comment by Lukeuke** [comment]
@liveans Hi! I could change it if it's not a problem. This would be my first and hopefully not last contribution.

**Comment by liveans** [comment]
> [@liveans](https://github.com/liveans) Hi! I could change it if it's not a problem. This would be my first and hopefully not last contribution.

Hello @Lukeuke, sure. If you have any questions regarding to the problem, you can always reach out, happy to help.
You might want to take a look at this, [runtime - CONTRIBUTING](https://github.com/dotnet/runtime/blob/main/CONTRIBUTING.md).

**Comment by Lukeuke** [comment]
Great, I've started writing tests but I see the issue. Im getting ``System.InvalidOperationException : Cannot re-call BeginGetRequestStream/BeginGetResponse while a previous call is still in progress.`` Is this expected behaviour?

```csharp
[Fact]
public async Task GetRequestStreamAsync_CalledTwice_ReturnsSameStream()
{
    await LoopbackServer.CreateClientAndServerAsync(
        async uri =>
        {
            HttpWebRequest request = WebRequest.CreateHttp(uri);
            request.Method = "POST";

            // First call — write to the request stream
            using (StreamWriter writer = new StreamWriter(await request.GetRequestStreamAsync()))
            {
                writer.Write("Hello, world!");
            }

            // Second call — close stream again
            (await request.GetRequestStreamAsync()).Close();

            // Should succeed — if request stream handling is correct
            using WebResponse response = await request.GetResponseAsync();
        },
        async server =>
        {
            await server.AcceptConnectionAsync(async connection =>
            {
                await connection.ReadRequestHeaderAsync();

                using StreamWriter reader = new StreamReader(connection.Stream, leaveOpen: true);
                string body = await reader.ReadToEndAsync();

                Assert.Contains("Hello", body);

                await connection.SendResponseAsync();
            });
        });
}
```

**Comment by liveans** [comment]
> Great, I've started writing tests but I see the issue. Im getting `System.InvalidOperationException : Cannot re-call BeginGetRequestStream/BeginGetResponse while a previous call is still in progress.` Is this expected behaviour?
> 
> [Fact]
> public async Task GetRequestStreamAsync_CalledTwice_ReturnsSameStream()
> {
>     await LoopbackServer.CreateClientAndServerAsync(
>         async uri =>
>         {
>             HttpWebRequest request = WebRequest.CreateHttp(uri);
>             request.Method = "POST";
> 
>             // First call — write to the request stream
>             using (StreamWriter writer = new StreamWriter(await request.GetRequestStreamAsync()))
>             {
>                 writer.Write("Hello, world!");
>             }
> 
>             // Second call — close stream again
>             (await request.GetRequestStreamAsync()).Close();
> 
>             // Should succeed — if request stream handling is correct
>             using WebResponse response = await request.GetResponseAsync();
>         },
>         async server =>
>         {
>             await server.AcceptConnectionAsync(async connection =>
>             {
>                 await connection.ReadRequestHeaderAsync();
> 
>                 using StreamWriter reader = new StreamReader(connection.Stream, leaveOpen: true);
>                 string body = await reader.ReadToEndAsync();
> 
>                 Assert.Contains("Hello", body);
> 
>                 await connection.SendResponseAsync();
>             });
>         });
> }

https://github.com/dotnet/runtime/blob/39a31f082e77fb8893016c30c0858f0e5f8c89ea/src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L1144-L1147

Seems like, we're setting `_beginGetRequestStreamCalled` to true, but never set it back to false, and exception message is a bit inconsistent with the name of the field.

Couple of thoughts for here:

- Either we can check both `_beginGetRequestStreamCalled` and `_endGetRequestStreamCalled` together to check if a call to the `BeginGetRequestStream` is in progress.
- Or we can set `_beginGetRequestStreamCalled` back to false at the end of the `EndGetRequestStream` and set `_endGetRequestStreamCalled` back to false at the start of `BeginGetRequestStream`.

.NET Framework also allows this behavior with async code as well, and it also returns the same stream.

These changes should be safe IMO, from a quick check. Sync call doesn't have this restriction anyway.

**Comment by Lukeuke** [comment]
@liveans I've opened a PR with sync test that i wrote but without use of LoopbackServer. 
This test below that I've tried was causing an request timeout exception

```csharp
  await LoopbackServer.CreateClientAndServerAsync(
      async uri =>
      {
          var request = WebRequest.CreateHttp(uri);
          request.Method = "POST";

          using (var writer = new StreamWriter(request.GetRequestStream()))
              writer.Write("Hello, world!");
          request.GetRequestStream().Close();

          request.GetResponse();

          await Task.CompletedTask;
      },
      async server =>
      {
          await server.AcceptConnectionAsync(async connection =>
          {
              await connection.ReadRequestHeaderAsync();

              await connection.SendResponseAsync();
          });
      });
```
